docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+If the comparison result is not-equal, the entire value is a `Float32`.
+If the comparison result is equal, then a value of bit 20 is tested, with zero indicating a `U16` and one indicating a `U20`, either of which can be extracted by a bit shift (note: On Arm systems, either value can be extracted with a single `UBFX` instruction, on x86 systems that support the BMI1 instructions a BEXTR can do the same).
+
+### Example: `Float64 | T* | U*`
+
+The payload size of `Float64` is 64 bits and it requires one discriminator state.
+The payload size of `T*` and `U*` is each either 32, 48, or 128 bits, depending on the target architecture and they require one discriminator state between them.
+The lowering for this union therefore depends on the target.
+
+On 32-bit systems, the largest type is `Float64` and the remaining types all fit easily in the significand bits.
+There are no discriminator bits required but the 'as-if' rules triggers as if there were a single bit for determining which end of the space should be used and the remaining values are stored at the end of the space that is strongly aligned: bits 0-31.
+
+On 64-bit systems, the same rules apply, packing the used bits of the pointer into bits 0-50.
+
+On CHERI systems, the pointer types are larger than the `Float64` and so the payload is a 128-bit value, distinguished by the tag bit.
+
+### Example: `T* | None`
+
+In this example, `None` is a singleton type with no fields, used as a placeholder indicating a not-present value.
+The `T*` requires one pointer's worth of space for the payload (as above, this varies depending on the target) and one discriminator state.
+The `None` value requires one discriminator state and no payload.
+
+The largest value is a pointer and so on CHERI systems the tag bit is used as the discriminator, on other architectures the low bit of the pointer is used (because the pointer is the largest type and the discriminator fits in a three-bit value).
+The discriminator values are assigned starting from the singletons, so zero is used for `None` and one for `T*`.
+
+A pattern match expression on this type is compiled into a compare-and-branch-on-{non}-zero, which is a single instruction on most architectures.
+Accessing the `T*` on a CHERI system requires no further operations.
+On integer-pointer systems, it requires subtracting one.
+On any mainstream architecture other than RISC-V, this subtraction can be folded into the immediate offset on a load instruction and so requires no additional instructions.
+
+If the `None` value is used (rather than just being pattern matched on and ignored) then the compiler must be able to materialise the address (or capability) to the singleton.
+This is no worse than if we stored the address inline; however, because in that case the match would require a comparison against the address and so the address would need to be loaded before the pattern match, rather than conditionally in one path.
```

Comment:
This seems orthogonal to the type encoding, as you'd have to do that anyway, even if the descriptor was a whole new byte.

Reply:
I don't think I understand this comment.  The alternative representation is to use the address of a `None` singleton global as the descriptor and just match on the descriptor field in the pointee object, rather than special casing for 'is the pointer 0?'

Reply:
I can't remember what I meant by this and I also don't understand my own question... :S ignore me.